https://www.youtube.com/playlist?list=PLqnslRFeH2UrcDBWF5mfPGpqQDSta6VK4

In [1]:
import torch

In [2]:
x = torch.empty(1)
x

tensor([-3.1470e-24])

In [3]:
x = torch.empty(2, 3)
x

tensor([[ 5.3113e-26,  3.0850e-41, -3.1470e-24],
        [ 4.5898e-41,  8.9683e-44,  0.0000e+00]])

In [4]:
x = torch.empty(2, 2, 3)
x

tensor([[[5.2303e-26, 3.0850e-41, 0.0000e+00],
         [0.0000e+00, 5.2306e-26, 3.0850e-41]],

        [[5.2306e-26, 3.0850e-41, 5.2306e-26],
         [3.0850e-41, 8.4078e-45, 0.0000e+00]]])

In [5]:
x = torch.rand(2, 2)
x

tensor([[0.7800, 0.5788],
        [0.1137, 0.6791]])

In [ ]:
x = torch.zeros(2, 2)
x

In [6]:
x = torch.ones(2, 2)
x

tensor([[1., 1.],
        [1., 1.]])

In [7]:
x = torch.ones(2, 2, dtype=torch.uint8)
x

tensor([[1, 1],
        [1, 1]], dtype=torch.uint8)

In [8]:
x = torch.ones(2, 2, dtype=torch.float16)
x

tensor([[1., 1.],
        [1., 1.]], dtype=torch.float16)

In [9]:
x = torch.ones(2, 2, dtype=torch.float)
print(x)
type(x)

tensor([[1., 1.],
        [1., 1.]])


torch.Tensor

In [10]:
x = torch.rand(2, 2)
y = torch.rand(2, 2)

print(x, y)

z = x + y
print(z)

tensor([[0.3823, 0.2984],
        [0.5039, 0.9981]]) tensor([[0.1281, 0.5000],
        [0.5810, 0.6475]])
tensor([[0.5104, 0.7983],
        [1.0848, 1.6456]])


In [11]:
import numpy as np

a = np.ones(5)
print(a)

b = torch.from_numpy(a)
print(b)

[1. 1. 1. 1. 1.]
tensor([1., 1., 1., 1., 1.], dtype=torch.float64)


In [12]:
if torch.cuda.is_available():
    device = torch.device("cuda")
    x = torch.ones(5, device=device)
    y = torch.ones(5)
    y = y.to(device)
    z = x + y
    print(z)
    z = z.to("cpu")
    print(z)

tensor([2., 2., 2., 2., 2.], device='cuda:0')
tensor([2., 2., 2., 2., 2.])


In [13]:
x = torch.ones(5, requires_grad=True)
print(x)

tensor([1., 1., 1., 1., 1.], requires_grad=True)


만약 .requires_grad 속성을 True 로 설정하면, 그 tensor에서 이뤄진 모든 연산들을 추적(track)하기 시작합니다.   
계산이 완료된 후 .backward() 를 호출하여 모든 변화도(gradient)를 자동으로 계산할 수 있습니다.   
이 Tensor의 변화도는 .grad 속성에 누적됩니다.   
https://pytorch.org/docs/stable/notes/autograd.html

## Autograd

In [14]:
import torch

w = torch.tensor(2.0, requires_grad=True)
# 값이 2인 임의 스칼라텐서 w
# requires_grad=True: 이 텐서에 대한 기울기를 저장함

y = w**2
z = 2*y + 5

z.backward()

print('dz / dw : {}'.format(w.grad))

dz / dw : 8.0


In [24]:
x = torch.tensor(2.0, requires_grad=True)
print(x)

y = 2*x 
print(y)

z = 3*y
print(z)

w = 3+x
print(w)

z.backward()

print(x.grad)
print(type(x.grad))

print('dz / dx : {}'.format(x.grad))

tensor(2., requires_grad=True)
tensor(4., grad_fn=<MulBackward0>)
tensor(12., grad_fn=<MulBackward0>)
tensor(5., grad_fn=<AddBackward0>)
tensor(6.)
<class 'torch.Tensor'>
dz / dx : 6.0


In [26]:
# PyTorch Tutorial 03 - Gradient Calculation With Autograd
import torch

weights = torch.ones(4, requires_grad=True)

for epoch in range(3):
    model_output = (weights*3).sum()
    
    model_output.backward()
    
    print(weights.grad)
    
    weights.grad.zero_()
    # optimizer.zero_grad 랑 비슷한 듯

tensor([3., 3., 3., 3.])
tensor([3., 3., 3., 3.])
tensor([3., 3., 3., 3.])


In [27]:
weights = torch.ones(4, requires_grad=True)

for epoch in range(3):
    model_output = (weights*3).sum()
    
    model_output.backward()
    
    print(weights.grad)
    # 변화도가 누적됨

tensor([3., 3., 3., 3.])
tensor([6., 6., 6., 6.])
tensor([9., 9., 9., 9.])


In [ ]:
import torch

x = torch.tensor(1.0)
y = torch.tensor(2.0)

w = torch.tensor(1.0, requires_grad=True)

# forward pass and compute the loss

y_hat = w * x
loss = (y_hat - y)**2

print(loss)

# backward pass
loss.backward()
print(w.grad)

# update weights
# next forward and backward

---

In [21]:
import torch

x = torch.tensor((1.0, 2.0, 3.0), requires_grad=True)
print(x)

tensor([1., 2., 3.], requires_grad=True)


In [22]:
y = x + 2
print(y)

tensor([3., 4., 5.], grad_fn=<AddBackward0>)


In [11]:
# A = torch.Tensor([2.2571, 2.2923, 1.8443])
# B = torch.Tensor([2.2571, 2.2923, 1.8443])
# print(2.2571*2.2571*2)
# print(A*B*2)

In [27]:
z = y*y*2 # 2y^2
print(z)
z = z.mean() # RuntimeError: grad can be implicitly created only for scalar outputs
# print(z)

tensor([18., 32., 50.], grad_fn=<MulBackward0>)


In [28]:
z.backward() # dz/dx
print(x.grad)

tensor([ 8.0000, 10.6667, 13.3333])


결과 숫자가 무슨뜻인지 모르겠음

---

PyTorch의 모든 신경망의 중심에는 **autograd** 패키지가 있습니다. 먼저 이것을 가볍게 살펴본 뒤, 첫번째 신경망을 학습시켜보겠습니다.

autograd 패키지는 Tensor의 모든 연산에 대해 **자동 미분**을 제공합니다. 이는 실행-기반-정의(define-by-run) 프레임워크로, 이는 코드를 어떻게 작성하여 실행하느냐에 따라 역전파가 정의된다는 뜻이며, 역전파는 학습 과정의 매 단계마다 달라집니다.

도함수를 계산하기 위해서는 Tensor 의 **.backward()** 를 호출하면 됩니다. 만약 Tensor 가 스칼라(scalar)인 경우(예. 하나의 요소 값만 갖는 등)에는 backward 에 인자를 정해줄 필요가 없습니다. 하지만 여러 개의 요소를 갖고 있을 때는 tensor의 모양을 gradient 의 인자로 지정할 필요가 있습니다.



In [44]:
type(torch.autograd.backward)
# type(torch.autograd.backward(torch.tensor(1.)))

function

In [38]:
import torch

x = torch.ones(2,2, requires_grad=True)
print(x)

tensor([[1., 1.],
        [1., 1.]], requires_grad=True)


In [39]:
y = x + 2
print(y)

tensor([[3., 3.],
        [3., 3.]], grad_fn=<AddBackward0>)


y는 연산의 결과로 생성된 것이므로 grad_fn을 갖는다

In [40]:
print(y.grad_fn)

In [41]:
z = y * y * 3
out = z.mean()

print(z, out)

tensor([[27., 27.],
        [27., 27.]], grad_fn=<MulBackward0>) tensor(27., grad_fn=<MeanBackward0>)


.requires_grad_( ... ) 는 기존 Tensor의 requires_grad 값을 바꿔치기 (in-place)하여 변경합니다. 입력값이 지정되지 않으면 기본값은 False 입니다.

In [45]:
a = torch.randn(2, 2)
a = ((a * 3) / (a - 1))

print(a.requires_grad)
a.requires_grad_(True)
print(a.requires_grad)

False
True


In [46]:
b = (a * a).sum()
print(b.grad_fn)

In [47]:
print(z, out)

tensor([[27., 27.],
        [27., 27.]], grad_fn=<MulBackward0>) tensor(27., grad_fn=<MeanBackward0>)


In [48]:
out.backward()
print(x.grad) # 변화도 d(out)/dx 를 출력

tensor([[4.5000, 4.5000],
        [4.5000, 4.5000]])


mean 때문에 1/4이므로

3*2*(1+2) / 4 = 18 / 4 = 4.5

일반적으로, torch.autograd 는 벡터-야코비안 곱을 계산하는 엔진입니다.